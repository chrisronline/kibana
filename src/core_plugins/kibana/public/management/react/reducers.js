/* eslint-disable */
import { combineReducers } from 'redux';
import { routerReducer } from 'react-router-redux'

import app from './app/app.reducer';
import indexPattern from './index-pattern/index-pattern.reducer';

export const rootReducer = combineReducers({
  routing: routerReducer,
  app,
  indexPattern,
});
