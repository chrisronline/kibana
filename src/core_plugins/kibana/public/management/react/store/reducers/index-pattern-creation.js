import { handleActions } from 'redux-actions';
import { createSelector } from 'reselect';
import { chunk, sortBy as sortByLodash } from 'lodash';

import {
  selectTimeField,
  fetchedTimeFields,
  goToNextPage,
  goToPreviousPage,
  changeSort,
  fetchedIndices,
  includeSystemIndices,
  excludeSystemIndices,
  creatingIndexPattern,
  createdIndexPattern,
} from '../actions/index-pattern-creation';

import {
  getIndexPatternCreate,
} from '../../reducers';

const defaultState = {
  isIncludingSystemIndices: false,
  isCreating: false,
  timeFields: {
    timeFields: undefined,
    selectedTimeField: undefined,
  },
  results: {
    indices: undefined,
    pattern: undefined,
    hasExactMatches: false,
    perPage: 10,
    page: 1,
    sortBy: undefined,
    sortAsc: true,
  }
}

export default handleActions({
  [selectTimeField](state, { payload }) {
    const { timeFields } = payload;

    return {
      ...state,
      timeFields: {
        ...state.timeFields,
        timeFields,
      },
    };
  },
  [fetchedTimeFields](state, { payload }) {
    const { timeFields } = payload;

    return {
      ...state,
      timeFields: {
        ...state.timeFields,
        timeFields,
      },
    };
  },
  [selectTimeField](state, { payload }) {
    const { timeField } = payload;

    return {
      ...state,
      timeFields: {
        ...state.timeFields,
        selectedTimeField: timeField,
      },
    };
  },
  [fetchedIndices](state, { payload }) {
    const { indices, hasExactMatches, pattern } = payload;

    return {
      ...state,
      results: {
        ...state.results,
        indices,
        hasExactMatches,
        pattern,
      },
    };
  },
  [changeSort](state, { payload }) {
    const { sortBy, sortAsc } = payload;

    return {
      ...state,
      results: {
        ...state.results,
        sortBy,
        sortAsc: sortBy === state.results.sortBy ? !state.results.sortAsc: sortAsc,
      },
    };
  },
  [goToNextPage](state, { payload }) {
    return {
      ...state,
      results: {
        ...state.results,
        page: getNextPage(state.results),
      },
    };
  },
  [goToPreviousPage](state, { payload }) {
    return {
      ...state,
      results: {
        ...state.results,
        page: getPreviousPage(state.results),
      },
    };
  },
  [includeSystemIndices](state, action) {
    return {
      ...state,
      isIncludingSystemIndices: true,
    };
  },
  [excludeSystemIndices](state, action) {
    return {
      ...state,
      isIncludingSystemIndices: false,
    };
  },
  [creatingIndexPattern](state, action) {
    return {
      ...state,
      isCreating: true,
    };
  },
  [createdIndexPattern](state, action) {
    return {
      ...state,
      isCreating: false,
    };
  },
}, defaultState);

function getNextPage({ page, perPage, indices }) {
  const pages = Math.ceil(indices.length / perPage);
  const nextPage = page + 1;
  return nextPage > pages
    ? 1
    : nextPage;
}

function getPreviousPage({ page, perPage, indices }) {
  const pages = Math.ceil(indices.length / perPage);
  const previousPage = page - 1;
  return previousPage < 1
    ? pages
    : previousPage;
}

function getFilteredIndices(indices, isIncludingSystemIndices) {
  if (!indices) {
    return indices;
  }

  if (isIncludingSystemIndices) {
    return indices;
  }

  // All system indices begin with a period.
  return indices.filter(index => !index.name.startsWith('.'));
}

function getPaginatedIndices(indices, page, perPage) {
  const pagesOfIndices = chunk(indices, perPage);
  return pagesOfIndices[page - 1];
}

export const getFilteredAndPaginatedIndices = createSelector(
  [
    state => state.results,
    state => state.isIncludingSystemIndices
  ],
  ({
    indices: allIndices,
    page,
    perPage,
    sortBy,
    sortAsc,
  }, isIncludingSystemIndices) => {
    if (allIndices === undefined) {
      return {
        indices: undefined,
        numOfPages: 0,
      };
    }

    let filteredIndices = getFilteredIndices(allIndices, isIncludingSystemIndices);
    if (!!sortBy) {
      filteredIndices = sortByLodash(filteredIndices, sortBy);
      if (!sortAsc) {
        filteredIndices.reverse();
      }
    }

    const numOfPages = Math.ceil(filteredIndices.length / perPage);
    const indices = getPaginatedIndices(filteredIndices, page, perPage);
    return {
      indices,
      numOfPages,
    }
  }
);

export const getPattern = state => getIndexPatternCreate(state).results.pattern;
export const getSelectedTimeField = state => getIndexPatternCreate(state).timeFields.selectedTimeField;
