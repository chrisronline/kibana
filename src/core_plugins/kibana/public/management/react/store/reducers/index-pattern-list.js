import { handleActions } from 'redux-actions';

import {
  fetchedIndexPatterns,
  setTransientTableId,
} from '../actions/index-pattern-list';

import {
  getIndexPatternList,
} from '../../reducers';

const defaultState = {
  indexPatterns: undefined,
  transientTableId: undefined,
};

export default handleActions({
  [fetchedIndexPatterns](state, { payload }) {
    const { indexPatterns } = payload;

    return {
      ...state,
      indexPatterns,
    };
  },
  [setTransientTableId](state, { payload: { id: transientTableId } }) {
    return {
      ...state,
      transientTableId,
    };
  }
}, defaultState);

export const getIndexPatterns = state => getIndexPatternList(state).indexPatterns;
export const getPathToIndexPatterns = () => 'indexPatterns';
