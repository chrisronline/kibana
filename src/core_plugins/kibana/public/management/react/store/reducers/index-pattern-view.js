import { handleActions } from 'redux-actions';

import {
  fetchedIndexPattern,
  setAsDefaultIndexPattern,
} from '../actions/index-pattern-view';

import {
  getIndexPatternView,
} from '../../reducers';

const defaultState = {
  indexPattern: {
    pattern: undefined,
    fields: undefined,
    timeFieldName: undefined,
    isDefault: undefined,
    id: undefined,
  },
  fieldsTable: {},
  tabs: {},
};

export const indexPatternView = handleActions({
  [fetchedIndexPattern](state, { payload: { indexPattern } }) {
    return {
      ...state,
      indexPattern,
    };
  },
  [setAsDefaultIndexPattern](state) {
    return {
      ...state,
      indexPattern: {
        ...state.indexPattern,
        isDefault: true,
      }
    };
  }
}, defaultState);

export const getPathToFields = () => 'indexPattern.fields';
export const getTabs = state => getIndexPatternView(state).tabs;
