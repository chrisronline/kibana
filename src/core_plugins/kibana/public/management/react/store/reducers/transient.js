import { handleActions } from 'redux-actions';

import {
  change,
  createItem,
} from '../actions/transient';

const defaultState = {};

export default handleActions({
  [createItem](state, { payload: { id, defaultData }}) {
    return {
      ...state,
      [id]: defaultData || {},
    }
  },
  [change](state, { payload: { id, data } }) {
    return {
      ...state,
      [id]: {
        ...state[id],
        ...data
      },
    };
  },
}, defaultState);
