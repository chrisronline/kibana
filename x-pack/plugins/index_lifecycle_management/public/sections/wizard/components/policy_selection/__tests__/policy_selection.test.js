/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */

import React from 'react';
import { Provider } from 'react-redux';
import { mount } from 'enzyme';
import { PolicySelection } from '../';
import { indexLifecycleManagement } from '../../../../../store/reducers';

jest.mock('@elastic/eui', () => ({
  EuiFlexGroup: props => props.children || '',
  EuiFlexItem: props => props.children || '',
  EuiButton: props => props.children || '',
  EuiDescribedFormGroup: props => props.children || '',
  EuiFormRow: props => props.children || '',
  EuiSelect: props => props.children || '',
}));

jest.mock('../../../../../store/actions', () => ({
  fetchPolicies: jest.fn(),
  setSelectedPolicy: jest.fn(),
}));

const defaultState = indexLifecycleManagement({}, {});
const defaultStore = {
  subscribe: jest.fn(),
  dispatch: jest.fn(),
};
const actions = require('../../../../../store/actions');

describe('PolicySelection', () => {
  it('should render properly', () => {
    const store = {
      getState: () => ({
        ...defaultState,
        indexTemplate: {
          ...defaultState.indexTemplate,
          indexTemplates: [{}]
        }
      }),
      ...defaultStore,
    };

    const component = mount(
      <Provider store={store}>
        <PolicySelection/>
      </Provider>
    );

    expect(component.find('PolicySelection')).toMatchSnapshot();
  });

  it('should call the selected policy prop when selecting a policy', () => {
    const policy = { name: 'foo' };
    const store = {
      getState: () => ({
        ...defaultState,
        policies: {
          ...defaultState.policies,
          policies: [
            policy
          ]
        }
      }),
      ...defaultStore,
    };

    const component = mount(
      <Provider store={store}>
        <PolicySelection/>
      </Provider>
    );

    component.find('PolicySelection').find('EuiSelect').get(0).props.onChange({ target: { value: policy.name } });
    expect(actions.setSelectedPolicy).toBeCalledWith(policy);
  });

  it('should call the selected policy prop when creating a new policy', () => {
    const policy = { name: 'foo' };
    const store = {
      getState: () => ({
        ...defaultState,
        policies: {
          ...defaultState.policies,
          policies: [
            policy
          ]
        }
      }),
      ...defaultStore,
    };

    const component = mount(
      <Provider store={store}>
        <PolicySelection/>
      </Provider>
    );

    component.find('PolicySelection').find('EuiButton').get(0).props.onClick();
    expect(actions.setSelectedPolicy).toBeCalledWith(undefined);
  });
});
