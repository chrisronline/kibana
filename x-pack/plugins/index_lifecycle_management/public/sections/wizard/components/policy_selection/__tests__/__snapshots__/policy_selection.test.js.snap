// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PolicySelection should render properly 1`] = `
<PolicySelection
  fetchPolicies={[Function]}
  policies={Array []}
  selectedPolicyName=""
  setSelectedPolicy={[Function]}
>
  <EuiDescribedFormGroup
    description="An index lifecycle policy is a blueprint for transitioning your data over time. You can create a new policy or edit an existing policy and save it with a new name."
    fullWidth={true}
    title={
      <h4>
        Select or create a policy
      </h4>
    }
    titleSize="s"
  >
    <EuiFlexGroup
      alignItems="center"
    >
      <EuiFlexItem>
        <EuiFormRow
          label="Existing policies"
        >
          <EuiSelect
            onChange={[Function]}
            options={
              Array [
                Object {
                  "text": "-- New Policy --",
                  "value": "",
                },
              ]
            }
            value=""
          />
        </EuiFormRow>
      </EuiFlexItem>
      <EuiFlexItem
        grow={false}
      >
        <EuiFormRow
          hasEmptyLabelSpace={true}
        >
          <EuiButton
            onClick={[Function]}
          >
            Create new policy
          </EuiButton>
        </EuiFormRow>
      </EuiFlexItem>
    </EuiFlexGroup>
  </EuiDescribedFormGroup>
</PolicySelection>
`;
